apps/autorelayer-interop/src/relayer.ts

```diff
@@ -294,28 +319,62 @@ export class Relayer {
    * @param messages - The pending messages to group
    * @returns A mapping of gas tank provider to pending messages
    */
-  private groupPendingMessagesByGasTankProvider(
+  private async groupPendingMessagesByGasTankProvider(
+    accounts: Address[],
     messages: PendingMessageWithGasTank[],
-  ): Map<GasTankProvider, { messages: PendingMessageWithGasTank[] }> {
-    return messages.reduce((acc, message) => {
+  ): Promise<Map<GasTankProvider, { messages: PendingMessageWithGasTank[] }>> {
+    // Cache balance calculations to avoid repeated calls
+    const balanceCache = new Map<GasTankProvider, number>()
+
+    /**
+     * Gets the cached balance for a gas tank provider
+     * @param provider - The gas tank provider
+     * @returns The cached balance
+     */
+    const getCachedBalance = async (
+      provider: GasTankProvider,
+    ): Promise<number> => {
+      if (!balanceCache.has(provider)) {
+        balanceCache.set(
+          provider,
+          await this.getGasProviderBalance(provider, accounts).catch(
+            (error) => {
+              this.log.error(
+                { err: error },
+                `failed to get gas provider balance ${provider.gasProviderAddress}`,
+              )
+              throw error
+            },
+          ),
+        )
+      }
+      return balanceCache.get(provider)!
+    }
+
+    const result = new Map<
+      GasTankProvider,
+      { messages: PendingMessageWithGasTank[] }
+    >()
+
+    for (const message of messages) {
+      let bestGasProvider = message.gasTankProviders[0]
+
       // Find the gas provider with highest balance for this message
-      const bestGasProvider = message.gasTankProviders.reduce(
-        (best, provider) =>
-          this.getGasProviderBalance(provider) >
-          this.getGasProviderBalance(best)
-            ? provider
-            : best,
-        message.gasTankProviders[0],
-      )
+      for (const provider of message.gasTankProviders) {
+        const balance = await getCachedBalance(provider)
+        if (balance > bestGasProvider.gasProviderBalance) {
+          bestGasProvider = provider
+        }
+      }
```

Comment:
🤔 Is it odd that the "bestGasProvider" may not have enough balance to relay the message? We could that the provider has enough balance to send the message, but it would make the response object a little more complicated. Maybe it's not worth worrying about?